crates/bevy_ecs/src/component.rs

```diff
@@ -261,7 +279,69 @@ impl ComponentInfo {
 
     /// Create a new [`ComponentInfo`].
     pub(crate) fn new(id: ComponentId, descriptor: ComponentDescriptor) -> Self {
-        ComponentInfo { id, descriptor }
+        ComponentInfo {
+            id,
+            descriptor,
+            hooks: ComponentHooks::default(),
+        }
+    }
+
+    /// Register a [`ComponentHook`] that will be run when this component is added to an entity
+    ///
+    /// Will panic if the component already has an `on_add` hook
+    pub fn on_add(&mut self, hook: ComponentHook) -> &mut Self {
+        assert!(
+            self.hooks.on_add.is_none(),
+            "Component id: {:?}, already has an on_add hook",
+            self.id()
+        );
+        self.hooks.on_add = Some(hook);
+        self
+    }
+
+    /// Register a [`ComponentHook`] that will be run when this component is added or set by `.insert`
+    ///
+    /// Will panic if the component already has an `on_insert` hook
+    pub fn on_insert(&mut self, hook: ComponentHook) -> &mut Self {
+        assert!(
+            self.hooks.on_insert.is_none(),
+            "Component id: {:?}, already has an on_insert hook",
+            self.id()
+        );
+        self.hooks.on_insert = Some(hook);
+        self
+    }
+
+    /// Register a [`ComponentHook`] that will be run when this component is removed from an entity.
```

Comment:
It mightn't be obvious to all users immediately that an entity being de-spawned does not count as its components being removed so wouldn't hurt to be explicit about this in the docs.

Reply:
Remove hooks do run on entity despawn but we should probably call that out specifically to make that clear as it's not the first time the question has been asked.

Reply:
Oh invert what I said then :smile: